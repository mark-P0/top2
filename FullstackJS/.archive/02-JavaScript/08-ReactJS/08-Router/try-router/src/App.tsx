import { BrowserRouter, Routes, Route } from "react-router-dom";
import routeTreeFn from "./routes";

export default function App() {
  const trees = Object.entries(routeTreeFn).map(([path, RouteTree], idx) => (
    <Route key={idx} path={path} element={<RouteTree />} />
  ));

  return (
    <BrowserRouter>
      <Routes>{trees}</Routes>
    </BrowserRouter>
  );
}
