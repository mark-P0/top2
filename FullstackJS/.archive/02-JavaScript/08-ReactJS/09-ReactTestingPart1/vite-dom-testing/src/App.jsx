export default function App() {
  return <h1>Our first test</h1>;
}
