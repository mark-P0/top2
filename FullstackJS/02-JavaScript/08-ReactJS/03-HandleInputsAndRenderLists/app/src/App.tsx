import "./tailwind.js";
import React from "react";
import ReactDOM from "react-dom/client";
import { Overview } from "./components/Overview.js";
import { CreateTaskEvent, DeleteTaskEvent } from "./events.js";

export class Task {
  static #ct = 0;

  id: number;
  description: string;
  constructor(description: string) {
    this.id = ++Task.#ct;
    this.description = description;
  }

  toString() {
    return this.description;
  }
}

class NewTaskForm extends React.Component<{}> {
  #createTask: React.FormEventHandler = (event) => {
    event.preventDefault();

    const form = event.target as HTMLFormElement;
    const data = new FormData(form);
    const description = data.get("description")?.toString();

    if (!description || description === "") return;

    CreateTaskEvent.publish(description);

    form.reset();
  };

  render() {
    return (
      <form className="bg-white rounded-lg flex gap-3 p-3" onSubmit={this.#createTask}>
        <input
          className="border border-2 border-stone-300 rounded-full flex-1 px-3"
          type="text"
          name="description"
        />
        <button className="border border-2 border-stone-300 rounded px-3 py-1" type="submit">
          Add Task
        </button>
      </form>
    );
  }
}

type Props = {};
type State = { tasks: Task[] };

class App extends React.Component<Props, State> {
  #CreateTaskSubscriber = async (description: string) => {
    const { tasks } = this.state;
    this.setState({
      tasks: [...tasks, new Task(description)],
    });
  };
  #DeleteTaskSubscriber = async (taskToDelete: Task) => {
    const { tasks } = this.state;
    this.setState({
      tasks: tasks.filter((task) => task !== taskToDelete),
    });
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      tasks: [],
    };
  }

  componentDidMount() {
    CreateTaskEvent.subscribe(this.#CreateTaskSubscriber);
    DeleteTaskEvent.subscribe(this.#DeleteTaskSubscriber);
  }

  componentWillUnmount() {
    CreateTaskEvent.unsubscribe(this.#CreateTaskSubscriber);
    DeleteTaskEvent.unsubscribe(this.#DeleteTaskSubscriber);
  }

  render() {
    const { tasks } = this.state;

    return (
      <div className="min-h-screen min-w-screen bg-stone-500 flex flex-col gap-3 p-3">
        <NewTaskForm />
        <Overview tasks={tasks} />
      </div>
    );
  }
}

ReactDOM.createRoot(document.getElementById("root") as HTMLElement).render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
