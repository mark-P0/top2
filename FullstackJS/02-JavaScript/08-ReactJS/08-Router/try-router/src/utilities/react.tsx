/**
 * Combine class string arguments into a single space-delimited string.
 * Allows for breaking down class strings into composable pieces.
 *
 * **Truthy/Falsy values are simply ignored.** They are accepted to
 * support the conditional shorthand pattern. Think of the packages
 * [`clsx`](https://www.npmjs.com/package/clsx) <!-- cspell:disable-line -->
 * and [`classnames`](https://www.npmjs.com/package/classnames).
 *
 * If supported, CSS services (e.g. Tailwind's VSCode extension)
 * can be configured to be enabled within calls to this helper.
 *
 * Example:
 * ```js
 * const classes = C(
 *  'cls1 cls2 cls3',
 *  'annotated classes', // A comment explaining this class bit
 *  C(
 *    'group',
 *    'related',
 *    // 'classes', // Classes can be easily disabled for testing
 *  ),
 *  // `&&` evaluates to the right class strings if the left condition
 *  // is truthy; otherwise, it will evaluate to `false`
 *  // and will be silently ignored.
 *  shouldUseTheseClasses && 'classes to be conditionally used'
 *  true && 'use these classes',
 *  false && 'do-not use classes',
 *  null && 'these classes will NOT be used',
 *  undefined && 'NEITHER will these classes',
 * )
 * ```
 */
export function C(...classes: (string | boolean | null | undefined)[]): string {
  /**
   * Type Predicate pattern
   *
   * https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates
   */
  const strings = classes.filter((cls): cls is string => typeof cls === 'string');
  return strings.join(' ');
}