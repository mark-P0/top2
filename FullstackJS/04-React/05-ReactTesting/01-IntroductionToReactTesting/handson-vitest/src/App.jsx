import { useState } from "react";

export function App1() {
  return <h1>Our First Test</h1>;
}

export function App2() {
  const [heading, setHeading] = useState("Magnificent Monkeys");

  function clickHandler() {
    setHeading("Radical Rhinos");
  }

  return (
    <>
      <button type="button" onClick={clickHandler}>
        Click Me
      </button>
      <h1>{heading}</h1>
    </>
  );
}
