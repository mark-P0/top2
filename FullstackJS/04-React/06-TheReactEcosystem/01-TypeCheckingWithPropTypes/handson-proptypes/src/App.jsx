import { bool as PTBool, string as PTString } from "prop-types";
import { useState } from "react";

function Header({ text, showLinks = false }) {
  return (
    <header>
      {showLinks && (
        <nav>
          <a href="https://vitejs.dev">Vite</a>
          <a href="https://react.dev">React</a>
        </nav>
      )}
      <h1>{text}</h1>
    </header>
  );
}

Header.propTypes = {
  text: PTString.isRequired,
  showLinks: PTBool,
};

Header.defaultProps = {
  showLinks: "",
};

function App() {
  const [count, setCount] = useState(0);

  return (
    <>
      <Header text="Vite + React" />
      <div className="card">
        <button onClick={() => setCount((count) => count + 1)}>
          count is {count}
        </button>
        <p>
          Edit <code>src/App.jsx</code> and save to test HMR
        </p>
      </div>
      <p className="read-the-docs">
        Click on the Vite and React logos to learn more
      </p>
    </>
  );
}

export default App;
