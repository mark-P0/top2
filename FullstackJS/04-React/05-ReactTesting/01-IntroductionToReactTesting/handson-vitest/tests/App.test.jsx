import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it } from "vitest";
import { App1, App2 } from "../src/App.jsx";

describe("First app", () => {
  it("renders headline", () => {
    render(<App1 />);

    const heading = screen.getByRole("heading").textContent;
    expect(heading).toMatch(/our first test/i);
  });
});

describe("Second app", () => {
  it("renders magnificent monkeys by default", () => {
    // since screen does not have the container property, we'll destructure render to obtain a container for this test
    const { container } = render(<App2 />);
    expect(container).toMatchSnapshot();
  });

  it("renders radical rhinos after button click", async () => {
    const user = userEvent.setup();
    render(<App2 />);

    const button = screen.getByRole("button", { name: "Click Me" });
    await user.click(button);

    const heading = screen.getByRole("heading").textContent;
    expect(heading).toMatch(/radical rhinos/i);
  });
});
