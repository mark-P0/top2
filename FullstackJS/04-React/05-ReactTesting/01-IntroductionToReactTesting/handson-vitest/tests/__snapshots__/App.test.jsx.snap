// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Second app > renders magnificent monkeys 1`] = `
<div>
  <button
    type="button"
  >
    Click Me
  </button>
  <h1>
    Magnificent Monkeys
  </h1>
</div>
`;

exports[`Second app > renders magnificent monkeys by default 1`] = `
<div>
  <button
    type="button"
  >
    Click Me
  </button>
  <h1>
    Magnificent Monkeys
  </h1>
</div>
`;
